File: benchmarks/tasks/account_state_evolution/starter.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Active:
    pass


@dataclass(frozen=True)
class Locked:
    failed_attempts: int


@dataclass(frozen=True)
class Closed:
    pass


type AccountState = Active | Locked | Closed


def can_sign_in(state: AccountState) -> bool:
    match state:
        case Active():
            return True
        case Locked():
            return False
        case Closed():
            return False


def status_label(state: AccountState) -> str:
    match state:
        case Active():
            return "active"
        case Locked(failed_attempts):
            return f"locked:{failed_attempts}"
        case Closed():
            return "closed"

File: benchmarks/tasks/account_state_evolution/starter.rs
#[derive(Debug, PartialEq, Eq)]
pub enum AccountState {
    Active,
    Locked(i64),
    Closed,
}

pub fn can_sign_in(state: AccountState) -> bool {
    match state {
        AccountState::Active => true,
        AccountState::Locked(_) => false,
        AccountState::Closed => false,
    }
}

pub fn status_label(state: AccountState) -> String {
    match state {
        AccountState::Active => "active".to_string(),
        AccountState::Locked(failed_attempts) => format!("locked:{failed_attempts}"),
        AccountState::Closed => "closed".to_string(),
    }
}
